src/index.js

```diff
@@ -1,12 +1,37 @@
-import React from 'react'
+import React, { Suspense } from 'react'
 import { render, cleanup, act } from 'react-testing-library'
 
 function TestHook({ callback, hookProps, children }) {
-  try {
```

Comment:
If you kept this block an just re-threw you could also grab any promises to store in a third result value of 'promise' or the like. That would make testing suspense much easier because then we could await that. At the very least inspecting the promise thrown would be useful in tests.

Reply:
(If you rethrow the exact same error it will preserve the original stack trace.)

Reply:
The intention of removing this and using a `Suspense` and `ErrorBoundary` wrapper was to protect this library from having to deal with the internals of how error handling and suspended rendering actually works.  The fact that a thrown promise is what triggers react to suspend is an implementation for React, rather than something we should be relying on.

Reply:
you're still setting error property though, right? Seems like you are somewhat depending on react in that manner. regardless - this *is* a react hooks testing library. very specific to how react works.

Reply:
Yes, `error` is being set, but it's coming from the exposed `componentDidCatch` API, so we don't care _how_ React gets it to that point, just deal with it when it gets there.  When we look at the `Suspense` equivalent, React does not expose this to us in any way, so if we want to handle it, we essentially need to recreate react's internal logic in order to treat it the same way.  If React changes the specifics on how it deals with promises, we would also need to change.

Reply:
But I am interested to see an example of when you would use a captured error that is not covered by `waitForNextUpdate` and `result.current`.  IF the use case is strong enough, then perhaps the additional maintenance burden is worth it.  I just don't see it yet.

Reply:
Currently I'm wrapping everything in suspense so I can track when it gets suspended. Suspense triggering is a very common part of my tests. That can still be done with the fallback prop. I won't need to look at the exact promise if `waitForNextUpdate()` works, but still need:

- the promise to be caught so it doesn't throw above
- a mechanism to see when it is called

Reply:
> if waitForNextUpdate() works

It should.  I'd rather fix holes where it doesn't than custom suspense handling.

> - the promise to be caught so it doesn't throw above

This solution would not end up throwing out of the `renderHook` function, if that's what you mean?

> - a mechanism to see when it is called

Can you unpack this a bit more?  Are you concerned about when an update was triggered by something other than the promise resolving, or something else?

Reply:
One path is that no promise is thrown from the hook - I want to test that it actually threw a promise, and THEN did something else

Reply:
I guess I'm questioning the value of having that much pwoer to inspect the internals of what's happening in your hook (i.e. white box test) vs. just asserting the output from the given inputs (i.e. black box test).

In my experience, white box test are very fragile to changes in the system and need constant updating as the internals shift around (e.g. you might change the condition for _when_ the promise is thrown, so now would need to move things around in your tests to match the change).  The black box tests are more likely to continue to work so, long as the contract of the inputs and outputs hasn't changes (otherwise the test _must be_ changed anyway in both testing styles).

In cases where you need more control over when and how the promise thrown, then you may be better off extracting it into some form of factory pattern and mocking that behaviour in the hook test, and testing those variations of the factory seperately to the hook itself.

As an escape hatch, if you _really_ need access to the promise itself, I think you could just catch it in the hool callback itself

```js
let promise
const { result } = renderHook(() => {
  promise = null // assuming you want to reset it between renders
  try {
    useSomething()
  } catch (e) {
    if (e.then) {
      promise = e
    }
  }
})

const promiseValue = await promise

expect(promiseValue).toBe('whatever')
```

I'm still unsure how that is more useful than

```js
const { result, waitForUpdate } = renderHook(() => useSomething())

await waitForUpdate

expect(result.current).toBe('whatever you expect after the prmise resolved')
```

As in both cases
* if the promise resolves, and the expectation is met, the the test passes
* if the promise resolves, and the expectation is not met, the test fails
   * in the first example all you would know the promise value changed
   * in the second example you would only know something went wrong to affect the output
* if the promise rejects, the test fails with the thrown error
* if the promise is not thrown, the test will fail
   * in the first example it would probably be about trying to `await` `null`
   * in the second example the test would timeout waiting for an update that never comes

Reply:
I'm also leaning away from explicitly handling supsense as it would get very difficult to manage if you had a hook throw a promise, and then once resolved throw another promise, etc. Suspense would keep that component suspending untill it eventually rendered something, whereas keeping track of each one and what they were up to could get complicated.

Reply:
I mean, I would argue that suspense is very much an explicit interface and not an implementation detail - as anyone using the hook that throws promises will need to define suspense boundaries themselves. It's not about *how* but *if*.

Yes it could throw many times - that should be up to the test writer to understand exactly what that behavior should be - or if it even cares at all. I agree in many cases you won't care, and you can simply waitForUpdate. But there are certainly cases where you will